routing/endpointregistry.go

```diff
@@ -13,32 +14,66 @@ const defaultLastSeenTimeout = 1 * time.Minute
 // used to perform better load balancing, fadeIn, etc.
 type Metrics interface {
 	DetectedTime() time.Time
+	SetDetected(detected time.Time)
+
+	LastSeen() time.Time
+	SetLastSeen(lastSeen time.Time)
+
 	InflightRequests() int64
+	IncInflightRequest()
+	DecInflightRequest()
 }
 
 type entry struct {
-	detected         time.Time
-	inflightRequests int64
+	detected         atomic.Value // time.Time
+	lastSeen         atomic.Value // time.Time
+	inflightRequests atomic.Int64
 }
 
 var _ Metrics = &entry{}
 
 func (e *entry) DetectedTime() time.Time {
-	return e.detected
+	return e.detected.Load().(time.Time)
+}
+
+func (e *entry) LastSeen() time.Time {
+	return e.lastSeen.Load().(time.Time)
 }
 
 func (e *entry) InflightRequests() int64 {
-	return e.inflightRequests
+	return e.inflightRequests.Load()
+}
+
+func (e *entry) IncInflightRequest() {
+	e.inflightRequests.Add(1)
+}
+
+func (e *entry) DecInflightRequest() {
+	e.inflightRequests.Add(-1)
+}
+
+func (e *entry) SetDetected(detected time.Time) {
+	e.detected.CompareAndSwap(nil, detected)
```

Comment:
Why do we need this, can it be nil?

Reply:
I think we can also just use .Store and let caller decide if they want to update the value if stored value is not zero.

Reply:
The most clear solution I see is to create two methods `SetDetected` which does not overwrite the value and `setDetectedForce` which overwrites.
However, I don't see use case for the second one except entry's constructor, so I will keep the second one private, it will be easy to export it if we will need to.